domain/info.go

```diff
@@ -0,0 +1,180 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package domain
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/ddl"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/printer"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+const (
+	// ServerInformationPath store server information such as IP, port and so on.
+	ServerInformationPath = "/tidb/server/info"
+	// keyOpDefaultRetryCnt is the default retry count for etcd store.
+	keyOpDefaultRetryCnt = 2
+	// keyOpDefaultTimeout is the default time out for etcd store.
+	keyOpDefaultTimeout = 1 * time.Second
+)
+
+// InfoSyncer stores server info to Etcd when when the tidb-server starts and delete when tidb-server shuts down.
+type InfoSyncer struct {
+	etcdCli        *clientv3.Client
+	info           *ServerInfo
+	serverInfoPath string
+}
+
+// ServerInfo is server static information.
+// It will not update when server running. So please only put static information in ServerInfo struct.
+type ServerInfo struct {
+	ServerVersionInfo
+	ID         string `json:"ddl_id"`
+	IP         string `json:"ip"`
+	Port       uint   `json:"listening_port"`
+	StatusPort uint   `json:"status_port"`
+	Lease      string `json:"lease"`
+}
+
+// ServerVersionInfo is the server version and git_hash.
+type ServerVersionInfo struct {
+	Version string `json:"version"`
+	GitHash string `json:"git_hash"`
+}
+
+// NewInfoSyncer return new InfoSyncer. It is exported for testing.
+func NewInfoSyncer(id string, etcdCli *clientv3.Client) *InfoSyncer {
+	return &InfoSyncer{
+		etcdCli:        etcdCli,
+		info:           getServerInfo(id),
+		serverInfoPath: fmt.Sprintf("%s/%s", ServerInformationPath, id),
+	}
+}
+
+func getServerInfo(id string) *ServerInfo {
+	cfg := config.GetGlobalConfig()
+	info := &ServerInfo{
+		ID:         id,
+		IP:         cfg.AdvertiseAddress,
+		Port:       cfg.Port,
+		StatusPort: cfg.Status.StatusPort,
+		Lease:      cfg.Lease,
+	}
+	info.Version = mysql.ServerVersion
+	info.GitHash = printer.TiDBGitHash
+	return info
+}
+
+//GetServerInfo gets self server static information.
+func (is *InfoSyncer) GetServerInfo() *ServerInfo {
+	return is.info
+}
+
+// GetServerInfoByID gets server static information from Etcd.
+func (is *InfoSyncer) GetServerInfoByID(ctx context.Context, id string) (*ServerInfo, error) {
+	if is.etcdCli == nil || id == is.info.ID {
+		return is.info, nil
+	}
+	key := fmt.Sprintf("%s/%s", ServerInformationPath, id)
+	infoMap, err := getInfo(ctx, is.etcdCli, key, keyOpDefaultRetryCnt, keyOpDefaultTimeout)
```

Comment:
Why need to retrive all info? Could we just get `/tidb/server/info/xxx` from etcd?

Reply:
There have not get all info, the `infoMap` will at most contain 1 info.
`allInfo, err := getInfo(ctx, is.etcdCli, ServerInformationPath, keyOpDefaultRetryCnt, keyOpDefaultTimeout, clientv3.WithPrefix())` will get all info.

Reply:
But you get infos from `/tidb/server/info`, and you just need the `/tidb/server/info/id` actually.

Reply:
Oh. Never mind, I see, the key is different with GetAllServerInfo.